app/src/main/java/com/android/bookswap/model/map/Geolocation.kt

```diff
@@ -0,0 +1,74 @@
+package com.android.bookswap.model.map
+
+import android.Manifest
+import android.annotation.SuppressLint
+import android.app.Activity
+import android.content.pm.PackageManager
+import android.os.Looper
+import androidx.compose.runtime.mutableDoubleStateOf
+import androidx.core.app.ActivityCompat
+import com.google.android.gms.location.*
+
+class Geolocation(private val activity: Activity) {
+
+  private val REQUEST_LOCATION_PERMISSION = 1
+  private lateinit var fusedLocationClient: FusedLocationProviderClient
+  val latitude = mutableDoubleStateOf(Double.NaN) // Change to MutableState
+  val longitude = mutableDoubleStateOf(Double.NaN)
+
+  // Location request settings
+  private val locationRequest: LocationRequest =
+      LocationRequest.create().apply {
+        interval = 10000 // Update interval in milliseconds
+        fastestInterval = 5000 // Fastest update interval in milliseconds
+        priority = LocationRequest.PRIORITY_HIGH_ACCURACY
+      }
+
+  private val locationCallback =
+      object : LocationCallback() {
+        override fun onLocationResult(p0: LocationResult) {
+          p0.lastLocation.let { location ->
```

Comment:
Potential Issue: According to the Android documentation, lastLocation may return null in certain situations (e.g., when location services are disabled, the device hasn't received a location fix, or the user denies permissions). This can lead to a NullPointerException